SubList function

charan2628 opened the pull request:
### Description
Support extracting subLists from a list
 
### Issues Resolved
Resolves #[5529]
 
### Check List
- [X ] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

graytaylor0 commented on data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/SubListExpressionFunction.java:
```diff
@@ -0,0 +1,57 @@
+		if (startIndex < 0 || startIndex >= sourceList.size()) {
+			throw new RuntimeException("subList() start index should be between 0 and list length (inclusive)");
+		}
+		if (endIndex < 0 || endIndex > sourceList.size()) {
```
Let's allow for endIndex to be -1, which will mean we set it to sourceList.size()

charan2628 replied:
Updated if endIndex is -1 setting it to sourceList.size()

graytaylor0 commented on data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/SubListExpressionFunction.java:
```diff
@@ -0,0 +1,57 @@
+		if (!(value instanceof List)) {
+			throw new RuntimeException(key + " is not of list type");
+		}
+		List sourceList = (List)value;
```
It's better to use 

```
List<?> sourceList = (List<?>)value;
```

charan2628 replied:
Updated

graytaylor0 commented on data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/SubListExpressionFunction.java:
```diff
@@ -0,0 +1,57 @@
+		List sourceList = (List)value;
+		int startIndex = (Integer)args.get(1), endIndex = (Integer)args.get(2);
+
+		if (startIndex < 0 || startIndex >= sourceList.size()) {
```
I think this should be

```
if (startIndex < 0 || startIndex > sourceList.size())
```

charan2628 replied:
We are using 0 based index and startIndex is inclusive so startIndex equals to sourceList.size() causes index out of bounds for subList function, that's why I used >= 
For endIndex I used > since it's exclusive

graytaylor0 commented on data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/SubListExpressionFunction.java:
```diff
@@ -0,0 +1,57 @@
+			throw new RuntimeException("subList() takes 3 arguments");
+		}
+		if (!(args.get(0) instanceof String)) {
+			throw new RuntimeException("subList() takes 1st argument as string type");
```
Rather than throwing RuntimeException for these invalid arguments, can use IllegalArgumentException

charan2628 replied:
Update to IllegalArgumentException

graytaylor0 reviewed (commented):
Thanks for making this contribution. I left a couple comments.

graytaylor0 commented:
Also the build is failing due to lack of test coverage

```
[ant:jacocoReport] Rule violated for bundle data-prepper-expression: instructions covered ratio is 0.9, but expected minimum is 1.0

> Task :data-prepper-expression:jacocoTestCoverageVerification FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':data-prepper-expression:jacocoTestCoverageVerification'.
> Rule violated for bundle data-prepper-expression: instructions covered ratio is 0.9, but expected minimum is 1.0
```

charan2628 commented on data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/SubListExpressionFunction.java:
```diff
@@ -0,0 +1,57 @@
+			if (i >= startIndex && i < endIndex) {
+				result.add(sourceList.get(i));
+			}
+		}
```
Updated

charan2628 commented:
> Also the build is failing due to lack of test coverage, You also need to sign your commit. You can do git commit --amend and modify the message to include the signature. For example, this is my signature
> 
> ```
> Signed-off-by: Taylor Gray <tylgry@amazon.com>
> ```
> 
> ```
> [ant:jacocoReport] Rule violated for bundle data-prepper-expression: instructions covered ratio is 0.9, but expected minimum is 1.0
> 
> > Task :data-prepper-expression:jacocoTestCoverageVerification FAILED
> 
> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> Execution failed for task ':data-prepper-expression:jacocoTestCoverageVerification'.
> > Rule violated for bundle data-prepper-expression: instructions covered ratio is 0.9, but expected minimum is 1.0
> ```

Added test cases for 100% coverage.

san81 commented on data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/SubListExpressionFunctionTest.java:
Please add one test case where the key is nested in the event object for example, if the event object look like 
```
{
    "level1": {
           "level2": [ 1, 2, 3, 4, 5]
    }
}
```
and 
```
assertThat(subListExpressionFunction.evaluate(List.of("/level1/level2", 1, 3), testEvent, testFunction), equalTo(List.of(2, 3)));
```

charan2628 replied:
Added please check #testWithValidArgumentsCase3

graytaylor0 commented on data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/SubListExpressionFunction.java:
```diff
@@ -0,0 +1,53 @@
+			throw new RuntimeException("subList() end index should be between 0 and list length or -1 for list length (exclusive)");
+		}
+		if (startIndex > endIndex) {
+			throw new RuntimeException("subList() start index should be less or equal to end index");
```
Nit: Less than or equal to

charan2628 replied:
Changed

graytaylor0 commented on data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/SubListExpressionFunction.java:
```diff
@@ -0,0 +1,53 @@
+		if (!(args.get(0) instanceof String)) {
+			throw new IllegalArgumentException("subList() takes 1st argument as string type");
+		}
+		if (!(args.get(1) instanceof Integer) || !(args.get(2) instanceof Integer)) {
```
Did you run data prepper end to end to test this function? I pulled your PR and it looks like Integers can't be passed as arguments to the functions due to the grammar. I think we will need these to be strings for now and we can throw illegal argument exception if we can't parse the string to int.

```
  - add_entries:
        entries:
          - key: "result"
            value_expression: 'subList(/my_list, "0", "1")'
```

graytaylor0 commented:
> > Also the build is failing due to lack of test coverage, You also need to sign your commit. You can do git commit --amend and modify the message to include the signature. For example, this is my signature
> > ```
> > Signed-off-by: Taylor Gray <tylgry@amazon.com>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > [ant:jacocoReport] Rule violated for bundle data-prepper-expression: instructions covered ratio is 0.9, but expected minimum is 1.0
> > 
> > > Task :data-prepper-expression:jacocoTestCoverageVerification FAILED
> > 
> > FAILURE: Build failed with an exception.
> > 
> > * What went wrong:
> > Execution failed for task ':data-prepper-expression:jacocoTestCoverageVerification'.
> > > Rule violated for bundle data-prepper-expression: instructions covered ratio is 0.9, but expected minimum is 1.0
> > ```
> 
> Added test cases for 100% coverage.

Thanks! You still need to amend the commit message with the DCO though

charan2628 commented:
> > > Also the build is failing due to lack of test coverage, You also need to sign your commit. You can do git commit --amend and modify the message to include the signature. For example, this is my signature
> > > ```
> > > Signed-off-by: Taylor Gray <tylgry@amazon.com>
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > ```
> > > [ant:jacocoReport] Rule violated for bundle data-prepper-expression: instructions covered ratio is 0.9, but expected minimum is 1.0
> > > 
> > > > Task :data-prepper-expression:jacocoTestCoverageVerification FAILED
> > > 
> > > FAILURE: Build failed with an exception.
> > > 
> > > * What went wrong:
> > > Execution failed for task ':data-prepper-expression:jacocoTestCoverageVerification'.
> > > > Rule violated for bundle data-prepper-expression: instructions covered ratio is 0.9, but expected minimum is 1.0
> > > ```
> > 
> > 
> > Added test cases for 100% coverage.
> 
> Thanks! You still need to amend the commit message with the DCO though

Signed now

graytaylor0 reviewed (dismissed):
Thanks this looks good! Will you creating a PR to the documentation website for this (https://github.com/opensearch-project/documentation-website/blob/main/_data-prepper/pipelines/functions.md)?

charan2628 commented:
> Thanks this looks good! Will you creating a PR to the documentation website for this (https://github.com/opensearch-project/documentation-website/blob/main/_data-prepper/pipelines/functions.md)?

I will @graytaylor0

graytaylor0 commented:
> > Thanks this looks good! Will you creating a PR to the documentation website for this (https://github.com/opensearch-project/documentation-website/blob/main/_data-prepper/pipelines/functions.md)?
> 
> I will @graytaylor0

Thanks!

charan2628 commented on data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/SubListExpressionFunction.java:
```diff
@@ -0,0 +1,58 @@
+		int startIndex, endIndex;
+		try {
+			startIndex = args.get(1) instanceof Integer ? (Integer) args.get(1) : Integer.parseInt((String)args.get(1));
+			endIndex = args.get(2) instanceof Integer ? (Integer) args.get(2) : Integer.parseInt((String)args.get(2));
```
@graytaylor0 Done the changes to include both integer and string.

graytaylor0 commented:
> @graytaylor0 sublist should support both integers and strings for 2nd and 3rd arguments?
> 
> ```
>  - add_entries:
>         entries:
>           - key: "result"
>             value_expression: 'subList(/my_list, "0", "1")'
> ```
> 
> ```
>  - add_entries:
>         entries:
>           - key: "result"
>             value_expression: 'subList(/my_list, 0, 1)'
> ```

We could just require Integer here with the change I had shared previously. But if you were able to test this end to end with both string and integers then that works too

graytaylor0 reviewed (approved):
Thanks for addressing all the changes. Were you able to run some tests end to end before we merge?

charan2628 commented:
1

> Thanks for addressing all the changes. Were you able to run some tests end to end before we merge?

![image](https://github.com/user-attachments/assets/9fbed8c8-3d17-4fbb-b533-d708e7e4b65b)

```
simple-sample-pipeline:
  workers: 2
  delay: "5000"
  source:
    file:
      path: /home/schar/workspace/opensource/data-prepper-sublist/release/archives/linux/build/install/opensearch-data-prepper-2.11.0-SNAPSHOT-linux-x64/pipelines/input.json
      format: json
      record_type: event
      codec:
        newline:
  processor:
    - parse_json:
        source: message
    - add_entries:
        entries:
          - key: "sublist1"
            value_expression: "subList(/my_list, 0, 2)"
          - key: "sublist2"
            value_expression: 'subList(/my_list, "0", "2")'
          - key: "sublist3"
            value_expression: 'subList(/my_list, "3", "-1")'
    - delete_entries:
        with_keys:
           - message
  sink:
    - stdout:
```

```
{"my_list": [ 0, 1, 2, 3, 4, 5, 6]}
{"my_list": [ 0, 1, 2, 3, 4, 5, 6]}
{"my_list": [ 0, 1, 2, 3, 4, 5, 6]}
{"my_list": [ 0, 1, 2, 3, 4, 5, 6]}
{"my_list": [ 0, 1, 2, 3, 4, 5, 6]}
{"my_list": [ 0, 1, 2, 3, 4, 5, 6]}
{"my_list": [ 0, 1, 2, 3, 4, 5, 6]}
```

@graytaylor0 Done changes to remove double quotes if we receive string as arguments
Tested using above pipeline.yaml for both integer and string arguments with the input.json working as expected.

I name the new function as subList, in the original issue it was sublist, I will change it if needs to be changed

graytaylor0 reviewed (approved):
Thanks for all of this!

san81 reviewed (approved):
Thank you for this contribution!

dlvenable commented:
@charan2628 , Thank you for this great contribution!

The unit tests are failing.

You can more easily run all the relevant tests locally using this command:

```
./gradlew -p data-prepper-expression clean build
```

This will allow you to quickly test these changes without building everything.

Below are the failing tests. Some appear to be based on restrictions for other functions which are not allowed to take in numbers.

```
GenericExpressionEvaluator_ConditionalIT > testGenericExpressionEvaluatorThrows(String, Event) > [8] hasTags(10), org.opensearch.dataprepper.model.event.JacksonEvent@8206075 FAILED
    org.opentest4j.AssertionFailedError: Expected java.lang.RuntimeException to be thrown, but nothing was thrown.

GenericExpressionEvaluator_ConditionalIT > testGenericExpressionEvaluatorThrows(String, Event) > [15] contains(1234, /strField), org.opensearch.dataprepper.model.event.JacksonEvent@300f43a3 FAILED
    org.opentest4j.AssertionFailedError: Expected java.lang.RuntimeException to be thrown, but nothing was thrown.

GenericExpressionEvaluator_ConditionalIT > testGenericExpressionEvaluatorThrows(String, Event) > [17] contains(/strField, 1234), org.opensearch.dataprepper.model.event.JacksonEvent@5c1e1d0f FAILED
    org.opentest4j.AssertionFailedError: Expected java.lang.RuntimeException to be thrown, but nothing was thrown.

GenericExpressionEvaluator_ConditionalIT > testGenericExpressionEvaluatorThrows(String, Event) > [29] getMetadata(10), org.opensearch.dataprepper.model.event.JacksonEvent@8206075 FAILED
    org.opentest4j.AssertionFailedError: Expected java.lang.RuntimeException to be thrown, but nothing was thrown.

GenericExpressionEvaluator_ConditionalIT > testGenericExpressionEvaluatorThrows(String, Event) > [32] cidrContains(/sourceIp,123), org.opensearch.dataprepper.model.event.JacksonEvent@54d9de9c FAILED
    org.opentest4j.AssertionFailedError: Expected java.lang.RuntimeException to be thrown, but nothing was thrown.

ParseTreeCoercionServiceTest > testCoerceTerminalNodeLengthFunctionWithInvalidArgument() FAILED
    org.opentest4j.AssertionFailedError: Expected java.lang.RuntimeException to be thrown, but nothing was thrown.

ParseTreeTest > testSubtractOperator() FAILED
    java.lang.AssertionError: 
    Expected: is an instance of org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser$ArithmeticUnaryExpressionContext
         but: Expected context "/status_code==-200" | ConditionalExpressionContext to match is an instance of org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser$ArithmeticUnaryExpressionContext
    		Expected context "/status_code==-200" | EqualityOperatorExpressionContext to match is an instance of org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser$ArithmeticUnaryExpressionContext
    		Expected context "-200" | RegexOperatorExpressionContext to match is an instance of org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser$ArithmeticUnaryExpressionContext

1067 tests completed, 7 failed
    		Expected context "-200" | RelationalOperatorExpressionContext to match is an instance of org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser$ArithmeticUnaryExpressionContext
    		Expected context "-200" | SetOperatorExpressionContext to match is an instance of org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser$ArithmeticUnaryExpressionContext
    		Expected context "-200" | UnaryOperatorExpressionContext to match is an instance of org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser$ArithmeticUnaryExpressionContext
    		is an instance of org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser$ArithmeticUnaryExpressionContext
    		finally <[185 161 155 142 124 60]> is a org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser$UnaryOperatorExpressionContext

```

dlvenable reviewed (changes requested):
Need tests to pass.

charan2628 commented:
> Need tests to pass.

Working on it

charan2628 commented:
@dlvenable functions are now accepting Integer arguments, can I remove test cases checking for invalid arguments which are of type integer?

charan2628 commented:
@dlvenable Since functions now accepting Integer arguments I removed those scenarios as part of invalid argument case in GenericExpressionEvaluator_ConditionalIT and added new valid case where subList function accepting integer argument. All tests are now passing.

graytaylor0 commented:
@dlvenable I rebased this and checked it out locally to test on since it has been some time.
